File: android/app/src/main/java/info/bagen/rust/plaoc/microService/ipc/IpcEvent.kt
package info.bagen.rust.plaoc.microService.ipc

import com.google.gson.JsonObject
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import info.bagen.rust.plaoc.microService.helper.toBase64
import info.bagen.rust.plaoc.microService.helper.toUtf8
import java.lang.reflect.Type

class IpcEvent(val name: String, override val data: Any, override val encoding: IPC_DATA_ENCODING) :
    IpcMessage(IPC_MESSAGE_TYPE.EVENT), JsonSerializer<IpcEvent>, DataWithEncoding {

    companion object {
        inline fun fromBinary(name: String, data: ByteArray) =
            IpcEvent(name, data, IPC_DATA_ENCODING.BINARY)

        inline fun fromBase64(name: String, data: ByteArray) =
            IpcEvent(name, data.toBase64(), IPC_DATA_ENCODING.BASE64)

        inline fun fromUtf8(name: String, data: ByteArray) =
            fromUtf8(name, data.toUtf8())

        inline fun fromUtf8(name: String, data: String) =
            IpcEvent(name, data, IPC_DATA_ENCODING.UTF8)
    }

    @delegate:Transient
    val binary by lazy {
        dataToBinary()
    }

    @delegate:Transient
    val text by lazy {
        dataToText()
    }

    @delegate:Transient
    val jsonAble by lazy {
        when (encoding) {
            IPC_DATA_ENCODING.BINARY -> fromBase64(
                name,
                (data as ByteArray),
            )
            else -> this
        }
    }

    override fun serialize(
        src: IpcEvent, typeOfSrc: Type, context: JsonSerializationContext
    ) = JsonObject().also { jsonObject ->
        with(src.jsonAble) {
            jsonObject.add("type", context.serialize(type))
            jsonObject.addProperty("name", name)
            jsonObject.addProperty("data", data as String)
            jsonObject.add("encoding", context.serialize(encoding))
        }
    }

}

File: desktop-dev/src/core/ipc/IpcEvent.ts
import { CacheGetter } from "../../helper/cacheGetter.ts";
import { simpleDecoder } from "../../helper/encoding.ts";
import { $dataToBinary, $dataToText, IPC_DATA_ENCODING, IPC_MESSAGE_TYPE, IpcMessage } from "./const.ts";

export class IpcEvent extends IpcMessage<IPC_MESSAGE_TYPE.EVENT> {
  constructor(readonly name: string, readonly data: string | Uint8Array, readonly encoding: IPC_DATA_ENCODING) {
    super(IPC_MESSAGE_TYPE.EVENT);
  }
  static fromBase64(name: string, data: Uint8Array) {
    return new IpcEvent(name, simpleDecoder(data, "base64"), IPC_DATA_ENCODING.BASE64);
  }
  static fromBinary(name: string, data: Uint8Array) {
    return new IpcEvent(name, data, IPC_DATA_ENCODING.BINARY);
  }
  static fromUtf8(name: string, data: Uint8Array) {
    return new IpcEvent(name, simpleDecoder(data, "utf8"), IPC_DATA_ENCODING.UTF8);
  }
  static fromText(name: string, data: string) {
    return new IpcEvent(name, data, IPC_DATA_ENCODING.UTF8);
  }
  #binary = new CacheGetter(() => $dataToBinary(this.data, this.encoding));

  get binary() {
    return this.#binary.value;
  }

  #text = new CacheGetter(() => $dataToText(this.data, this.encoding));
  get text() {
    return this.#text.value;
  }
  #jsonAble = new CacheGetter(() => {
    if (this.encoding === IPC_DATA_ENCODING.BINARY) {
      return IpcEvent.fromBase64(this.name, this.data as Uint8Array);
    }
    return this as IpcEvent;
  });
  get jsonAble() {
    return this.#jsonAble.value;
  }
  toJSON() {
    return { ...this.jsonAble };
  }
}
